src/include/miopen/env.hpp

```diff
@@ -31,127 +31,164 @@
-#define MIOPEN_DECLARE_ENV_VAR(x)                 \
-    struct x                                      \
-    {                                             \
-        static const char* value() { return #x; } \
-    };
+template <typename T>
+struct ParseEnvVal
+{
+};
 
-/*
- * Returns false if a feature-controlling environment variable is defined
- * and set to something which disables a feature.
- */
-inline bool IsEnvvarValueDisabled(const char* name)
+template <>
+struct ParseEnvVal<bool>
 {
-    // NOLINTNEXTLINE (concurrency-mt-unsafe)
-    const auto value_env_p = std::getenv(name);
-    if(value_env_p == nullptr)
-        return false;
-    else
+    static bool go(const char* vp)
     {
-        std::string value_env_str = value_env_p;
+        std::string value_env_str{vp};
+
         for(auto& c : value_env_str)
         {
             if(std::isalpha(c) != 0)
             {
                 c = std::tolower(static_cast<unsigned char>(c));
             }
         }
-        return (std::strcmp(value_env_str.c_str(), "disable") == 0 ||
-                std::strcmp(value_env_str.c_str(), "disabled") == 0 ||
-                std::strcmp(value_env_str.c_str(), "0") == 0 ||
-                std::strcmp(value_env_str.c_str(), "no") == 0 ||
-                std::strcmp(value_env_str.c_str(), "off") == 0 ||
-                std::strcmp(value_env_str.c_str(), "false") == 0);
-    }
-}
 
-inline bool IsEnvvarValueEnabled(const char* name)
-{
-    // NOLINTNEXTLINE (concurrency-mt-unsafe)
-    const auto value_env_p = std::getenv(name);
-    if(value_env_p == nullptr)
-        return false;
-    else
-    {
-        std::string value_env_str = value_env_p;
-        for(auto& c : value_env_str)
+        if(value_env_str == "disable" || value_env_str == "disabled" || value_env_str == "0" ||
+           value_env_str == "no" || value_env_str == "off" || value_env_str == "false")
         {
-            if(std::isalpha(c) != 0)
-            {
-                c = std::tolower(static_cast<unsigned char>(c));
-            }
+            return false;
+        }
+        else if(value_env_str == "enable" || value_env_str == "enabled" || value_env_str == "1" ||
+                value_env_str == "yes" || value_env_str == "on" || value_env_str == "true")
+        {
+            return true;
+        }
+        else
+        {
+            MIOPEN_THROW(miopenStatusInvalidValue, "Invalid value for env variable");
         }
-        return (std::strcmp(value_env_str.c_str(), "enable") == 0 ||
-                std::strcmp(value_env_str.c_str(), "enabled") == 0 ||
-                std::strcmp(value_env_str.c_str(), "1") == 0 ||
-                std::strcmp(value_env_str.c_str(), "yes") == 0 ||
-                std::strcmp(value_env_str.c_str(), "on") == 0 ||
-                std::strcmp(value_env_str.c_str(), "true") == 0);
+
+        return false; // shouldn't reach here
     }
-}
+};
+
+template <>
+struct ParseEnvVal<uint64_t>
```

Comment:
Let's document allowed formats. It resided in the original code.
```suggestion
// Supports hexadecimals (with leading "0x"), octals (if prefix is "0") and decimals (default).
// Returns 0 if environment variable is in wrong format (strtoull fails to parse the string).
template <>
struct ParseEnvVal<uint64_t>
```